tools/clang/lib/AST/HlslTypes.cpp

```diff
@@ -129,15 +122,38 @@ bool IsHLSLNumericUserDefinedType(clang::QualType type) {
   return false;
 }
 
+bool IsUserDefinedType(clang::QualType QT) {
+  const clang::Type *Ty = QT.getCanonicalType().getTypePtr();
+  if (const RecordType *RT = dyn_cast<RecordType>(Ty)) {
+    const RecordDecl *RD = RT->getDecl();
+    if (RD->isImplicit())
+      return false;
+    if (auto TD = dyn_cast<ClassTemplateSpecializationDecl>(RD))
+      if (TD->getSpecializedTemplate()->isImplicit())
+        return false;
+    return true;
+  }
+  return false;
+}
```

Comment:
I feel like we have a lot of these. To wit, we have a very similarly named function `IsUserDefinedRecordType` at around line 589 which uses a very different approach, but seems to have the same intent.

Can you explain the method used here and used previously? Maybe we can consolidate the UDT detection functions?